hail/src/main/scala/is/hail/expr/types/encoded/EStruct.scala

```diff
@@ -0,0 +1,188 @@
+import is.hail.asm4s._
+import is.hail.expr.types.BaseType
+import is.hail.expr.types.physical._
+import is.hail.expr.types.virtual._
+import is.hail.io.{InputBuffer, OutputBuffer}
+import is.hail.utils._
+
+final case class EField(name: String, typ: EType, index: Int) {
+  def pretty(sb: StringBuilder, indent: Int, compact: Boolean) {
+    if (compact) {
+      sb.append(prettyIdentifier(name))
+      sb.append(":")
+    } else {
+      sb.append(" " * indent)
+      sb.append(prettyIdentifier(name))
+      sb.append(": ")
+    }
+    typ.pretty(sb, indent, compact)
+  }
+}
+
+final case class EStruct(fields: IndexedSeq[EField], override val required: Boolean = false) extends EBaseStruct {
+  lazy val virtualType: TStruct = TStruct(fields.map(f => Field(f.name, f.typ.virtualType, f.index)), required)
+
+  assert(fields.zipWithIndex.forall { case (f, i) => f.index == i })
+
+  val types: Array[EType] = fields.map(_.typ).toArray
+  val fieldNames: Array[String] = fields.map(_.name).toArray
+  val fieldIdx: Map[String, Int] = fields.map(f => (f.name, f.index)).toMap
+  def hasField(name: String): Boolean = fieldIdx.contains(name)
+  def fieldType(name: String): EType = types(fieldIdx(name))
+  val missingIdx = new Array[Int](size)
+  val nMissing: Int = EBaseStruct.getMissingness(types, missingIdx)
+  val nMissingBytes = (nMissing + 7) >>> 3
+
+  if (!fieldNames.areDistinct()) {
+    val duplicates = fieldNames.duplicates()
+    fatal(s"cannot create struct with duplicate ${plural(duplicates.size, "field")}: " +
+      s"${fieldNames.map(prettyIdentifier).mkString(", ")}", fieldNames.duplicates())
+  }
+
+  override def _decodeCompatible(pt: PType): Boolean = {
+    if (!pt.isInstanceOf[PStruct])
+      false
+    else {
+      val ps = pt.asInstanceOf[PStruct]
+      ps.required == required &&
+        ps.size <= size &&
+        ps.fields.forall { f =>
+          hasField(f.name) && fieldType(f.name).decodeCompatible(f.typ)
+        }
+    }
+  }
+
+  override def _encodeCompatible(pt: PType): Boolean = {
+    if (!pt.isInstanceOf[PStruct])
+      false
+    else {
+      val ps = pt.asInstanceOf[PStruct]
+      ps.required == required &&
+        size <= ps.size &&
+        fields.forall { f =>
+          ps.hasField(f.name) && f.typ.encodeCompatible(ps.fieldType(f.name))
+        }
+    }
+  }
+
+  def _decodedPType(requestedType: Type): PType = requestedType match {
+    case t: TInterval =>
+      val repr = t.representation
+      val pointType = _decodedPType(repr).asInstanceOf[PStruct].fieldType("start")
+      PInterval(pointType, required)
+    case t: TLocus => PLocus(t.rg, required)
+    case t: TStruct =>
+      val pFields = t.fields.map { case Field(name, typ, idx) =>
+        val pt = fieldType(name).decodedPType(typ)
+        PField(name, pt, idx)
+      }
+      PStruct(pFields, required)
+  }
+
+  override def _buildEncoder(pt: PType, mb: MethodBuilder, v: Code[_], out: Code[OutputBuffer]): Code[Unit] = {
```

Comment:
these can be defined on EBaseStruct, yes? They're duplicated on EStruct and ETuple.

In fact, I'm pretty sure these classes don't even need to exist at all.

One concern: we've lost the EstimableEmitter stuff here, which is mostly fine since we now wrap each encode/decode call in a method, but large structs will still create big methods. We should probably wrap into groups of 128 fields or something. I'm happy to leave that to a subsequent PR, but that should be the next step (supporting wide schemas is totally integral).

Reply:
Not quite, I could abstract the bits that are duplicated. Here's the diff

```diff
diff --git a/tmp/estructencode.scala b/tmp/etupleencode.scala
index 7df49fd44..7cfeb7f42 100644
--- a/tmp/estructencode.scala
+++ b/tmp/etupleencode.scala
@@ -1,5 +1,5 @@
   def _buildEncoder(pt: PType, mb: MethodBuilder, v: Code[_], out: Code[OutputBuffer]): Code[Unit] = {
-    val ft = pt.asInstanceOf[PStruct]
+    val ft = pt.asInstanceOf[PBaseStruct]
     val addr = coerce[Long](v)
     val writeMissingBytes = if (ft.size == size) {
       out.writeBytes(addr, ft.nMissingBytes)
@@ -11,10 +11,9 @@
         var b = const(0)
         var k = 0
         while (k < 8 && j < size) {
-          val f = fields(j)
-          if (!f.typ.required) {
-            val i = ft.fieldIdx(f.name)
-            b = b | (ft.isFieldMissing(addr, i).toI << k)
+          val et = types(j)
+          if (!et.required) {
+            b = b | (ft.isFieldMissing(addr, j).toI << k)
             k += 1
           }
           j += 1
@@ -28,15 +27,11 @@
       c
     }
 
-    val writeFields = Code(fields.map { ef =>
-      val i = ft.fieldIdx(ef.name)
-      val pf = ft.fields(i)
-      val encodeField = ef.typ.buildEncoder(pf.typ, mb)
-      val v = Region.loadIRIntermediate(pf.typ)(ft.fieldOffset(addr, i))
-      ft.isFieldDefined(addr, i).mux(
-        encodeField(v, out),
-        Code._empty[Unit]
-      )
+    val fieldEncoders = types.zipWithIndex.map { case (et, i) => (et.buildEncoder(ft.types(i), mb), i) }
+    val writeFields = Code(fieldEncoders.map {
+      case (encodeField, i) => ft.isFieldDefined(addr, i).mux(
+        encodeField(Region.loadIRIntermediate(ft.types(i))(ft.fieldOffset(addr, i)), out),
+        Code._empty[Unit])
     }: _*)
 
     Code(writeMissingBytes, writeFields, Code._empty[Unit])
```